frontend/src/components/AppHeader.tsx

```diff
@@ -1,53 +1,57 @@
-import { useCallback, useContext, useEffect, useRef, useState } from "react";
-import { UserContext } from "@/lib/UserContext";
+import { useCallback, useEffect, useRef, useState } from "react";
 import Link from "next/link";
 import Image from "next/image";
-import AppNavigation from "./AppNavigation";
-
 import styles from "@/styles/Nav.module.css";
-
+import AppNavigation from "./AppNavigation";
+import { useUser } from "@/context/UserContext";
+import { useWeb3 } from "@/context/Web3Context";
 import { magic } from "@/lib/magic";
-import { getUserData } from "@/lib/utils";
 
 export default function AppHeader({}) {
   const [navbarOpen, setNavbarOpen] = useState(false);
-  const [user, setUser] = useContext(UserContext);
+  const { user, setUser } = useUser();
+  const { initializeWeb3 } = useWeb3();
+
+  // Reference to the navigation menu
   const navMenuRef = useRef(null);
 
-  function openWallet() {
-    magic.wallet.getInfo().then((walletInfo) => {
-      if (walletInfo?.walletType == "magic") {
-        // NOTE: this will only work if the user has connected via a
-        // magic wallet, not via browser wallet (e.g. MetaMask)
-        magic.wallet.showUI().catch((err) => console.error(err));
-      } else {
-        // for non-magic wallets, copy the full wallet address to the clipboard
-        navigator.clipboard
-          .writeText(user?.address)
-          .then((res) =>
-            alert(`ETH wallet address copied to clipboard: ${user?.address}`),
-          );
-      }
-    });
-  }
-
-  function disconnect() {
-    // disconnect from magic
-    magic.wallet.disconnect();
-
-    // clear the state
-    setUser({});
-  }
-
-  function loginWithConnect() {
-    magic.wallet
-      .connectWithUI()
-      .then((res) => {
-        getUserData().then((data) => setUser(data));
-      })
-      .catch((err) => console.error(err));
-  }
+  // Function to handle wallet opening
+  const openWallet = async () => {
+    try {
+      // Try to show the magic wallet UI
+      // This will only work if the user has connected via a magic wallet, not via browser wallet (e.g. MetaMask)
+      await magic.wallet.showUI();
```

Comment:
One thing you could do here is catch the case where a user logs out from the settings page that's available through the showUI() method.

The request will emit a `promiEvent` called `disconnect`. Catching it would look like this

```
  magic.wallet.showUI().on('disconnect', () => {
    console.log('user disconnected');
    magic.user.logout();
  });
```

<img width="467" alt="Screenshot 2023-05-23 at 9 09 13 AM" src="https://github.com/magiclabs/magic-nft-template/assets/34493689/40db1632-6117-442b-b71c-2ff9a084095e">

Reply:
@hcote completely missed this, thanks for the callout